content/tutorial-reliable-transaction-submission.md

```diff
@@ -104,23 +104,35 @@ On restart, or the determination of a new last validated ledger (pseudocode):
 For each persisted transaction without validated result:
     Query transaction by hash
     If (result appears in validated ledger)
-		Persist the final result
-		If (result code is tesSUCCESS)
-			Application may act based on successful transaction
-		Else
-			Application may act based on failure
-			Maybe resubmit with new LastLedgerSequence and Fee
-	Else if (LastLedgerSequence > newest validated ledger)
-        Wait for more ledgers to be validated
-	Else
-		If (server has contiguous ledger history up to and
-	        including the ledger identified by LastLedgerSequence)
-			The transaction failed
-			Submit a new transaction, if appropriate for application
-		Else
-			Repeat submission of original transaction
+      # Outcome is final
+      Persist the final result
+      If (result code is tesSUCCESS)
+        Application may act based on successful transaction
+      Else
+        Application may act based on failure
+        Maybe resubmit with new LastLedgerSequence and Fee
+
+    Else if (LastLedgerSequence > newest validated ledger)
+      # Outcome is not yet final
+      Wait for more ledgers to be validated
+
+    Else
+      If (server has contiguous ledger history up to and
+            including the ledger identified by LastLedgerSequence)
+        The transaction failed
+        Submit a new transaction, if appropriate for application
+      Else
+        # Outcome is final, but not known due to a ledger gap
+        Wait to acquire contiguous ledger history
 ```
 
+#### Ledger Gaps
+
+If your server does not have contiguous ledger history from when the transaction was originally submitted up to and including the ledger identified by LastLedgerSequence, you may not know the final outcome of the transaction. (If it was included in one of the ledger versions your server is missing, you do not know whether it succeeded or failed.)
```

Comment:
> If it was included in one of the ledger versions your server is missing, you do not know whether it succeeded or failed.

Suggest changing "it" to "the transaction" to be totally clear.

Also, I'm having trouble understanding how the second sentence is different than the first. The second sentence is clearer to me. (That might just be me.)

Reply:
The pseudo-code also uses "contiguous". We can probably change both, though.

I can try to rephrase this paragraph again. I was trying to convey the difference between the following 3 cases:

- If the transaction is in a validated ledger you have, you know its outcome (succeed or fail) even if you have a ledger gap, because you only need that one ledger that has the transaction in question.
- If you have _all_ the ledgers the transaction _could_ be in and it isn't in any of them, you know it failed.
- If you don't have a ledger with the transaction in it, but the transaction could be in a ledger you don't have, you don't know the outcome of the transaction.

Reply:
Got it. 

In the first case, where would you get the validated ledger? From s2 or the Data API or something like that?

Anyway, the way it makes sense to me is:

* You know the outcome of the transaction because:
    * You have all of the possible the ledgers.
    * You have the ledger with the transaction.

* You don't know the outcome of a transaction because:
    * You don't have all of the possible ledgers.

Reply:
The first case is meant to describe how you could have a gap in _the range the transaction could have been in_, but it doesn't matter because you have _the one ledger the transaction actually ended up in_. So we're still talking about your own server's ledger store.

And while the way it makes sense to you is not wrong, there's a small difference between "not having all the possible ledgers" and "not having the ledger with the transaction". It's the difference between proving a positive ("it's right here") versus proving a negative ("it's not _anywhere_").